runtime/compiler/trj9/env/SystemSegmentProvider.cpp

```diff
@@ -98,18 +93,41 @@ J9::SystemSegmentProvider::request(size_t requiredSize)
       newSegment.heapAlloc,
       newSegment.heapTop
       );
+
+   TR::reference_wrapper<J9MemorySegment> newSegmentRef = TR::ref(newSegment);
+
    try
       {
-      _systemSegments.push_back(TR::ref(newSegment));
+      _systemSegments.push_back(newSegmentRef);
       }
    catch (...)
       {
       _systemSegmentAllocator.release(newSegment);
       throw;
       }
    _systemBytesAllocated += systemSegmentSize;
-   _currentSystemSegment = TR::ref(newSegment);
-   return allocateNewSegment(roundedSize);
+
+   // Rounded size determines when the segment is released, see J9::SystemSegmentProvider::release
+   if (roundedSize <= _systemSegmentSize)
+      {
+      // We want to use the remaining space of _currentSystemSegment after updating it.
+      // Carve its remaining space into small segments and add them to the free segment list so that we can use it later
+      //
+      while (remaining(_currentSystemSegment) >= defaultSegmentSize())
+         {
+         _freeSegments.push_back( TR::ref(allocateNewSegment(defaultSegmentSize(), _currentSystemSegment) ) );
+         }
+
+      _currentSystemSegment = newSegmentRef;
+      }
+   else
+      {
+      // _currentSystemSegment should not point to any segment with space larger than _systemSegmentSize because
+      // such segment cannot be reused for other request and is to be released once its region is out of scope.
```

Comment:
I think this comment is too specific to a Region. While it is true that an allocation, made through a `TR::Region`, that is larger than `_systemSegmentSize` will be released to the OS when the region goes out of scope, it isn't the only time that this could happen. If a different construct used `TR::SegmentProvider`'s `request`/`release` APIs directly, then when it calls `release`, it will also result in a large allocation being released to the OS. As such I think the comment should be along the lines of "such a segment cannot be reused for other requests and will be released when the release method is invoked, eg when a TR::Region goes out of scope".